modules/CbChargeSOMDriver/chargesom/CbChargeSOM.cpp

```diff
@@ -0,0 +1,260 @@
+            return types::cb_board_support::CPState::C;
+        default:
+            return types::cb_board_support::CPState::D;
+        }
+    }
+
+    if (voltage >= 2000 /* mV */) /* 2 V <= x < 4 V */
+        return types::cb_board_support::CPState::D;
+
+    if (voltage >= 1000 /* mV */) { /* 1 V <= x < 2 V */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+        case types::cb_board_support::CPState::C:
+        case types::cb_board_support::CPState::D:
+            return types::cb_board_support::CPState::D;
+        default:
+            return types::cb_board_support::CPState::E;
+        }
+    }
+
+    if (voltage >= -1000 /* mV */) /* -1 V <= x < 1 V */
+        return types::cb_board_support::CPState::E;
+
+    if (voltage >= -2000 /* mV */) { /* -2 V <= x < -1 V (not standard) */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+        case types::cb_board_support::CPState::C:
+        case types::cb_board_support::CPState::D:
+        case types::cb_board_support::CPState::E:
+            return types::cb_board_support::CPState::E;
+        default:
+            return types::cb_board_support::CPState::PilotFault;
+        }
+    }
+
+    if (voltage >= -10000 /* mV */) /* -10 V <= x < -2 V */
+        return types::cb_board_support::CPState::PilotFault;
+
+    if (voltage >= -11000 /* mV */) { /* -11 V <= x < -10 V (not standard) */
+        return previous_state == types::cb_board_support::CPState::F ? types::cb_board_support::CPState::F
+                                                                     : types::cb_board_support::CPState::PilotFault;
+    }
+
+    if (voltage >= -13000 /* mV */) /* -13 V <= x < -11 V */
+        return types::cb_board_support::CPState::F;
+
+    return types::cb_board_support::CPState::PilotFault; /* < -13 V */
+}
+
+double CbChargeSOM::cp_get_duty_cycle() const {
+    // this method must be called with `sc_mutex`already hold
+
+    return this->sc.duty_cycle;
+}
+
+void CbChargeSOM::cp_set_duty_cycle(double duty_cycle) {
+    std::scoped_lock lock(this->sc_mutex);
+
+    this->sc.duty_cycle = floorl(duty_cycle);
+
+    this->sync_with_hw();
+}
+
+bool CbChargeSOM::cp_is_nominal_duty_cycle() const {
+    // this method must be called with `sc_mutex`already hold
+
+    return 0 < this->sc.duty_cycle && this->sc.duty_cycle < 100;
+}
+
+bool CbChargeSOM::cp_is_enabled() {
+    std::scoped_lock lock(this->sc_mutex);
+
+    // FIXME
+    return true;
+}
+
+void CbChargeSOM::cp_disable() {
+    std::scoped_lock lock(this->sc_mutex);
+
+    // FIXME it is unclear atm if the current safety controller fw can disable CP
+    // so let's signal 100% for now
+    this->sc.duty_cycle = 100;
+
+    this->sync_with_hw();
+}
+
+types::board_support_common::Ampacity CbChargeSOM::get_ampacity(int& voltage) {
```

Comment:
Maybe we should mark this as untested? Or is it?

Reply:
It is "fully" tested in that I looked at the values which see using the Metrel and then manually tweaked the thresholds to reasonable values. Should work for now, but I not backed with "academic background".